bench-tps/src/bench.rs

```diff
@@ -1004,63 +1031,70 @@ fn fund_move_keys<T: Client>(
     info!("done funding keys, took {} ms", funding_time.as_ms());
 }
 
-pub fn generate_and_fund_keypairs<T: Client>(
-    client: &T,
+pub fn generate_and_fund_keypairs<T: 'static + Client + Send + Sync>(
+    client: Arc<T>,
     faucet_addr: Option<SocketAddr>,
     funding_key: &Keypair,
     keypair_count: usize,
     lamports_per_account: u64,
     use_move: bool,
-) -> Result<(Vec<Keypair>, Option<LibraKeys>, u64)> {
+) -> Result<(Vec<Keypair>, Option<LibraKeys>)> {
     info!("Creating {} keypairs...", keypair_count);
     let (mut keypairs, extra) = generate_keypairs(funding_key, keypair_count as u64);
     info!("Get lamports...");
 
-    // Sample the first keypair, see if it has lamports, if so then resume.
-    // This logic is to prevent lamport loss on repeated solana-bench-tps executions
-    let last_keypair_balance = client
-        .get_balance(&keypairs[keypair_count - 1].pubkey())
-        .unwrap_or(0);
+    // Sample the first keypair, to prevent lamport loss on repeated solana-bench-tps executions
+    let first_key = keypairs[0].pubkey();
+    let first_keypair_balance = client.get_balance(&first_key).unwrap_or(0);
+
+    // Sample the last keypair, to check if funding was already completed
+    let last_key = keypairs[keypair_count - 1].pubkey();
+    let last_keypair_balance = client.get_balance(&last_key).unwrap_or(0);
 
     #[cfg(feature = "move")]
     let mut move_keypairs_ret = None;
 
     #[cfg(not(feature = "move"))]
     let move_keypairs_ret = None;
 
-    if lamports_per_account > last_keypair_balance {
-        let (_blockhash, fee_calculator) = get_recent_blockhash(client);
-        let account_desired_balance =
-            lamports_per_account - last_keypair_balance + fee_calculator.max_lamports_per_signature;
-        let extra_fees = extra * fee_calculator.max_lamports_per_signature;
-        let mut total = account_desired_balance * (1 + keypairs.len() as u64) + extra_fees;
+    // 100 lamports should give enough wiggle room to handle source / dest keypair sets getting unbalanced
+    let minimum_balance = 100;
+
+    if first_keypair_balance < minimum_balance || last_keypair_balance < minimum_balance {
```

Comment:
I'm not sure I get this minimum balance check fixed at 100. The user could have specified 1000 or 10,000 lamports per account, then if they have only 100 they would not be funded, is that correct? The cluster fees can be in the 1000 lamport range also, so I'm not sure a fixed low value here will work in all cases.

Reply:
@sakridge you're right, this was oversimplified and didn't take tx fees into account. I updated the PR with a new approach: https://github.com/solana-labs/solana/pull/7723/commits/78516ca48db91cdbcfe79b3b314c11e8ecd96178